tests/integration/high_availability/test_replication.py

```diff
@@ -0,0 +1,251 @@
+from tests.integration.high_availability.high_availability_helpers import (
+    deploy_and_scale_mysql,
+    get_max_written_value_in_database,
+    high_availability_test_setup,
+)
+
+logger = logging.getLogger(__name__)
+
+TIMEOUT = 15 * 60
+
+
+@pytest.mark.order(1)
+@pytest.mark.abort_on_fail
+@pytest.mark.replication_tests
+async def test_kill_primary_check_reelection(ops_test: OpsTest) -> None:
+    """Test to kill the primary under load and ensure re-election of primary."""
+    mysql_application_name, _ = await high_availability_test_setup(ops_test)
+
+    mysql_unit = ops_test.model.applications[mysql_application_name].units[0]
+    primary = await get_primary_unit(ops_test, mysql_unit, mysql_application_name)
+    primary_name = primary.name
+
+    last_written_value = await get_max_written_value_in_database(ops_test, primary)
+
+    # kill the primary pod
+    client = lightkube.Client()
+    client.delete(Pod, primary.name.replace("/", "-"), namespace=ops_test.model.info.name)
+
+    async with ops_test.fast_forward():
+        # wait for model to stabilize, k8s will re-create the killed pod
+        await ops_test.model.wait_for_idle(
+            apps=[mysql_application_name],
+            status="active",
+            raise_on_blocked=True,
+            timeout=TIMEOUT,
+        )
+
+        # ensure a new primary was elected
+        mysql_unit = ops_test.model.applications[mysql_application_name].units[0]
+        new_primary = await get_primary_unit(ops_test, mysql_unit, mysql_application_name)
+        new_primary_name = new_primary.name
+
+        assert primary_name != new_primary_name
+
+        # wait (and retry) until the killed pod is back online in the mysql cluster
+        try:
+            for attempt in Retrying(stop=stop_after_delay(5 * 60), wait=wait_fixed(10)):
+                with attempt:
+                    cluster_status = await get_cluster_status(ops_test, mysql_unit)
+                    online_members = [
+                        label
+                        for label, member in cluster_status["defaultreplicaset"][
+                            "topology"
+                        ].items()
+                        if member["status"] == "online"
+                    ]
+                    assert len(online_members) == 3
+                    break
+        except RetryError:
+            assert False, "Old primary has not come back online after being killed"
+
+    # For 2 minutes, ensure that data is being written in the database by the
+    # continuous writes application charm
+    for attempt in Retrying(stop=stop_after_delay(2 * 60), wait=wait_fixed(3)):
+        with attempt:
+            for unit in ops_test.model.applications[mysql_application_name].units:
+                written_value = await get_max_written_value_in_database(ops_test, unit)
+                assert written_value > last_written_value, "Continuous writes not incrementing"
+
+                last_written_value = written_value
+
+
```

Comment:
verify old primary is up-to date after coming back, ie has all the same writes as the rest of the replicas

Reply:
actually, i inadvertently check that all units in the cluster have increasing writes [here](https://github.com/canonical/mysql-k8s-operator/pull/101/files#diff-b4dad27cf592fea29193749e4ff1dea90ffa41a7949e5b4c3eef94a056528c6aR87)

Reply:
addressed in [13bba1d](https://github.com/canonical/mysql-k8s-operator/pull/101/commits/13bba1d559a7066915acfc8042bad36b76e81ce8)